File: dsa/quicksort.c
#include <stdio.h>
void quickSort(int a[], int low, int high){
    int i,j,pivot,t;
    if (low <= high){
         i=low;
         j=high;
        pivot=low;
        
        while(i<j){
             while(a[i]<=a[pivot]){
                 i++;
             }
             while(a[j]>a[pivot]){
                 j--;
             }
             if(i<j){
                 t=a[i];
                 a[i]=a[j];
                 a[j]=t;
             }
             else{
                 t=a[j];
                 a[j]=a[pivot];
                 a[pivot]=t;
             }
            quickSort(a,low, j-1);
            quickSort(a, j+1,high);
        }
}
}
int main()
{
    int n, i;
    printf("enter array size : ");
    scanf("%d", &n);
    int a[n];
    printf("enter array : \n");
    for (i = 0; i < n; i++)
    {
         scanf("%d",&a[i]);
    }

    quickSort(a,0,n - 1);

    printf("Sorted array: \n");
    
    for (int i = 0; i < n; i++)
    {
        printf("%d ", a[i]);
         printf("\n");
    }
    return 0;
    }

File: REVISION/Divide and Conquer/QuickSort.java
import java.util.*;

class Sort {
    void Sort(int[] a, int low, int high) {
        if (low <= high) {
            int i = low+1;
            int j = high;
            int pivot = low;

            while (i < j) {
                while (i <= high && a[i] <= a[pivot]) {
                    i++;
                }
                while (a[j] > a[pivot]) {
                    j--;
                }
                if (i < j) {
                    int temp = a[i];
                    a[i] = a[j];
                    a[j] = temp;
                } else {
                    int temp = a[j];
                    a[j] = a[pivot];
                    a[pivot] = temp;
                }
                Sort(a, low, j - 1);
                Sort(a, j + 1, high);
            }
        }
    }
}
public class QuickSort{
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        System.out.print("Enter array size: ");
        int n = in.nextInt();
        int[] a = new int[n];
        Sort s = new Sort();

        System.out.println("Enter array:");
        for (int i = 0; i < n; i++) {
            a[i] = in.nextInt();
        }

        s.Sort(a, 0, n - 1);

        System.out.println("Sorted array:");
        for (int i = 0; i < n; i++) {
            System.out.println(a[i]);
        }

        in.close();
    }
}
